File: 20221207_Wednesday/20221207.go
// Leetcode: 938. Range Sum of BST.
// https://leetcode.com/problems/range-sum-of-bst/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 159 ms, faster than 58.46% of Go online submissions for Range Sum of BST.
// Memory Usage: 9.8 MB, less than 5.38% of Go online submissions for Range Sum of BST.
// 2022.12.07 Daily Challenge.
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rangeSumBST(root *TreeNode, low int, high int) int {
  if nil == root { return 0}
  r := root
  if low <= r.Val && r.Val <= high {
    sum := rangeSumBST(r.Left,low,high)+rangeSumBST(r.Right,low,high)
    return r.Val + sum
  } else if r.Val < low {
    return rangeSumBST(r.Right,low,high)
  } else if high < r.Val {
    return rangeSumBST(r.Left, low, high)
  }
  return -1
}

File: 20221207_Wednesday/20221207.rb
# Leetcode: 938. Range Sum of BST.
# https://leetcode.com/problems/range-sum-of-bst/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 232 ms, faster than 92.16% of Ruby online submissions for Range Sum of BST.
# Memory Usage: 216.6 MB, less than 66.67% of Ruby online submissions for Range Sum of BST.
# 2022.12.07 Daily Challenge.
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} low
# @param {Integer} high
# @return {Integer}
def range_sum_bst(root, low, high)
  return 0 if ! root
  x = root.val
  if x.between?(low,high)
    return x + range_sum_bst(root.left,low,high) + range_sum_bst(root.right, low, high)
  elsif x < low
    return range_sum_bst(root.right, low, high)
  elsif high < x
    return range_sum_bst(root.left, low, high)
  else
    raise "Unknown position."
  end
end

